keps/0027-force-pod-reload-parameters.md

```diff
@@ -0,0 +1,138 @@
+---
+kep-number: 27
+short-desc: A parameter flag to disable forced pod reload
+title: Forced Pod reload attribute
+authors:
+  - @aneumann82
+owners:
+  - @aneumann82
+editor:
+creation-date: 2020-03-31
+last-updated: 2020-03-31
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+## Summary
+
+This KEP describes addition of a flag for parameters that controls the forced pod reloading
+
+## Motivation
+
+The new flag will allow an operator developer to control that a specific attribute does not automatically restart
+all pods of a deployment or stateful set if this parameter is changed.
+
+### Goals
+
+Make it possible for an operator developer to specify that a parameter will *not* automatically restart all pods from
+a stateful set or a deployment.
+
+### Non-Goals
+
+- Detailed control when Pods are restarted: If the spec template for a stateful set is changed, Kubernetes will still
+reload all pods automatically.
+- Calculation of used config maps, secrets or other resources used by a pod: The automatic pod restart is currently 
+used to force pods to re-read config maps on a parameter update. When a user explicitly disables the pod restart for 
+a parameter, and this parameter updates a config map or another dependency of the pod, the change might not be applied
+in the pod until the next restart.
+
+## Proposal
+
+Add an additional attribute `forcePodRestart` to parameter specifications in `params.yaml`:
+
+```yaml
+  - name: NODE_COUNT
+    description: "Number of Cassandra nodes."
+    default: "3"
+    forcePodRestart: "false"
+```
+
+The default value for this parameter would be `true` to keep backwards compatibility. The general behavior should stay
+the way it is right now, that on a change of a parameter the pods will be automatically restarted.
+
+If multiple parameters are changed in one update, the `forcePodRestart` flags of all attributes are `OR`ed: If at least one
+parameter has the `forcePodRestart` set to `true`, the pods will execute a rolling restart.
+
+### User Stories
+
+- [#1424](https://github.com/kudobuilder/kudo/issues/1424) Fine control over pod restarts
+
+The main use case for this extension are big stateful sets that are sensitive to pod restarts, for example Cassandra. 
+
+Especially while changing the size of a stateful set, the pod definition itself is not modified at all and a full
+restart of all pods will have a negative impact.
+
+### Implementation Details/Notes/Constraints
+
+The forced pod restart is currently implemented by setting a UID in the attributes of the `podTemplate`.
+
+If the added parameter `forcePodRestart` is set to false, this UID should not be set, therefore keeping the podTemplate
+unchanged. This would prevent a rolling restart of all pods.
+
+Generally, the enhancer should not modify any template specs, but only the top-level resource if the flag is disabled.
```

Comment:
Is this true? I mean, the flag is called `forcePodRestart`, so we might omit to set the new UID for nested _pod templates only_.

Reply:
Alternatively, we might generalize this flag to smth. like `updateSpecTemplates/updateNestedSpecs/updateNestedResources` and never update any nested resources. Naming is hard.

Reply:
Yeah... I'm not fully sure on this, but I thought about why we add the annotations to the template specs at all. I think that's a remnant of the Kustomize behavior, and I don't really see a reason why we actually want to have the annotations in the templates.

The problem is that plan/phase/step annotations could change as well, and trigger pod restarts. I'd rather not have them in the template specs at all.

With regards to the different naming: I'm not sure the names you gave would help users much. Having `forcePodRestart` is a pretty speaking name, the other ones not so much. It would require quite a bit of understanding of KUDO and K8s internals to guess what they actually do.